File: python/66-PlusOne/main.py
from typing import List

import unittest

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        result = [0] * (len(digits) + 1)

        carry = 0
        input = len(digits) - 1
        output = len(result) - 1
        add = 1

        while input != -1:
            sum = digits[input] + add + carry
            result[output] = sum % 10
            carry = sum // 10

            input -= 1
            output -= 1
            add = 0

        if carry != 0:
            result[output] = 1

        if result[0] == 0:
            return result[1:]
        else:
            return result
        
class Tests(unittest.TestCase):
    def test_example1(self):
        solution = Solution()
        self.assertEqual(solution.plusOne([1,2,3]), [1,2,4])

    def test_example2(self):
        solution = Solution()
        self.assertEqual(solution.plusOne([4,3,2,1]), [4,3,2,2])

    def test_example3(self):
        solution = Solution()
        self.assertEqual(solution.plusOne([9,9]), [1,0,0])


if __name__ == '__main__':
    unittest.main()

File: go/66-PlusOne/main.go
package main

import "fmt"

func plusOne(digits []int) []int {
	result := make([]int, len(digits)+1)

	carry := 1

	for i := len(digits) - 1; i >= 0; i-- {
		total := carry + digits[i]

		result[i+1] = total % 10
		carry = total / 10
	}

	if carry == 0 {
		return result[1:]
	}

	result[0] = carry
	return result
}

func main() {
	fmt.Println(plusOne([]int{1, 2, 3}))
	fmt.Println(plusOne([]int{9, 9}))
}
